server/app/lib/actions/fusor/deployment/rhev/ose_launch.rb

```diff
@@ -75,6 +96,74 @@ def ose_launch_completed
+            else
+              Utils::Fusor::CommandUtils.run_command("mkdir -p #{get_ssh_key_path}")
+              Utils::Fusor::CommandUtils.run_command("ssh-keygen -f #{get_ssh_key_path}/id_#{key_type} -N '' -t #{key_type}")
+              ::Fusor.log.debug "====== ssh keys have been generated"
+            end
+          end
+
+          def host_exists?(hostname)
+            system("ping -c1 -w1 #{hostname}")
+          end
+
+          def wait_for_host(hostname)
+            ::Fusor.log.info "====== waiting for #{hostname} started..."
+            max_retry = 180
+            wait = 10
+            count = 0
+            while count < max_retry do
+              if host_exists?(hostname)
+                sleep(60)
+                ::Fusor.log.debug "====== #{hostname} is available!"
+                return
+              else
+                ::Fusor.log.debug "====== waiting for #{hostname} ====== (#{count}/#{max_retry})"
+                sleep(wait)
+                count += wait
+              end
+            end
+            ::Fusor.log.error "====== waiting for #{hostname} - TIMED OUT!!!"
+          end
+
+          def get_ssh_key_path
+            deployment = ::Fusor::Deployment.find(input[:deployment_id])
+            File.join(Rails.root, 'ssh-keys', 'openshift', "#{deployment.label}-#{deployment.id}")
+          end
+
+          def copy_keys_to_root(hostname, key_type, password)
+            ::Fusor.log.info "copying file to #{hostname}."
+
+            client = Utils::Fusor::SSHConnection.new(hostname, 'root', password)
+            client.execute("mkdir -p ~/.ssh && chmod 700 ~/.ssh")
+
+            ssh_keys = ["id_#{key_type}", "id_#{key_type}.pub"]
+            ssh_keys.each { |key|
+              from_path = "#{get_ssh_key_path}/#{key}"
+              to_path   = "/root/.ssh/#{key}"
+              Net::SCP.start(hostname, "root", :password => password, :paranoid => false) do |scp|
+                scp.upload!(from_path, to_path)
+              end
+            }
+          end
+
+          def copy_keys_to_user(username, hostname, key_type, password)
+            client = Utils::Fusor::SSHConnection.new(hostname, 'root', password)
+            client.execute("useradd #{username}")
+            client.execute("runuser -l #{username} -c 'mkdir ~/.ssh && chmod 700 ~/.ssh'")
+            client.execute("cp ~/.ssh/id_#{key_type}.pub /home/#{username}/.ssh/")
+            client.execute("chown #{username} /home/#{username}/.ssh/id_#{key_type}.pub")
+            client.execute("chgrp #{username} /home/#{username}/.ssh/id_#{key_type}.pub")
+            client.execute("runuser -l #{username} -c 'more ~/.ssh/id_#{key_type}.pub > ~/.ssh/authorized_keys'")
+            client.execute("runuser -l #{username} -c 'chmod 644 ~/.ssh/id_#{key_type}.pub'")
+            client.execute("runuser -l #{username} -c 'chmod 644 ~/.ssh/authorized_keys'")
+          end
```

Comment:
I agree that we should search for alternative ways to do this after this iteration.